precompiles/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+			[0x53, 0x44, 0x0c, 0x90] => {
+				return Self::is_contributor(&input[SELECTOR_SIZE_BYTES..]);
+			}
+			[0x76, 0xf7, 0x02, 0x49] => {
+				return Self::reward_info(&input[SELECTOR_SIZE_BYTES..]);
+			}
+			[0x4e, 0x71, 0xd9, 0x2d] => Self::claim()?,
+			_ => {
+				log::trace!(
+					target: "crowdloan-rewards-precompile",
+					"Failed to match function selector in crowdloan rewards precompile"
+				);
+				return Err(ExitError::Other(
+					"No crowdloan rewards wrapper method at given selector".into(),
+				));
+			}
+		};
+
+		let outer_call: Runtime::Call = inner_call.into();
+		let info = outer_call.get_dispatch_info();
+
+		// Make sure enough gas
+		if let Some(gas_limit) = target_gas {
+			let required_gas = Runtime::GasWeightMapping::weight_to_gas(info.weight);
+			if required_gas > gas_limit {
+				return Err(ExitError::OutOfGas);
+			}
+		}
+		log::trace!(target: "crowdloan-rewards-precompile", "Made it past gas check");
+
+		// Dispatch that call
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+
+		log::trace!(target: "crowdloan-rewards-precompile", "Gonna call with origin {:?}", origin);
+
+		match outer_call.dispatch(Some(origin).into()) {
+			Ok(post_info) => {
+				let gas_used = Runtime::GasWeightMapping::weight_to_gas(
+					post_info.actual_weight.unwrap_or(info.weight),
+				);
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Stopped,
+					cost: gas_used,
+					output: Default::default(),
+					logs: Default::default(),
+				})
+			}
+			Err(e) => {
+				log::trace!(
+					target: "crowdloan-rewards-precompile",
+					"Crowdloan rewards call via evm failed {:?}",
+					e
+				);
+				Err(ExitError::Other(
+					"Crowdloan rewards call via EVM failed".into(),
+				))
+			}
+		}
+	}
+}
+
+impl<Runtime> CrowdloanRewardsWrapper<Runtime>
+where
+	Runtime: pallet_crowdloan_rewards::Config + pallet_evm::Config + frame_system::Config,
+	Runtime::AccountId: From<H160>,
+	BalanceOf<Runtime>: TryFrom<U256> + TryInto<u128> + Debug,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	Runtime::Call: From<pallet_crowdloan_rewards::Call<Runtime>>,
+{
+	// The accessors are first. They directly return their result.
+	fn is_contributor(input: &[u8]) -> Result<PrecompileOutput, ExitError> {
+		// parse the address
+		let contributor = H160::from_slice(&input[12..32]);
+
+		log::trace!(
+			target: "crowdloan-rewards-precompile",
+			"Checking whether {:?} is a contributor",
+			contributor
+		);
+
+		let account: Runtime::AccountId = contributor.into();
+		// fetch data from pallet
+		let is_contributor =
+			pallet_crowdloan_rewards::Pallet::<Runtime>::accounts_payable(account).is_some();
+
+		log::trace!(target: "crowldoan-rewards-precompile", "Result from pallet is {:?}", is_contributor);
+
+		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
```

Comment:
Ideally we calculate the gas requirements and bail early before any actual I/O.

Reply:
Updated

Reply:
I'm seeing that `let gas_consumed = ...` is duplicated in this fn now (or is GH doing something weird because of this conversation?)